In [ ]:
from sklearn.datasets import load_wine
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder
import pandas as pd
import numpy as np
from sklearn.cluster import AgglomerativeClustering,KMeans
from sklearn.metrics import silhouette_score
from sklearn.metrics.cluster import adjusted_rand_score
import warnings
warnings.filterwarnings("ignore")


wine = load_wine()

X,y = wine.data,wine.target

In [ ]:
labels = AgglomerativeClustering(n_clusters=3,linkage='single',metric='euclidean').fit_predict(X)
silhouette = silhouette_score(X,labels)
print(f"Distancia: euclidean, Linkage:single, Silhouette: {silhouette}")

Distancia: euclidean, Linkage:single, Silhouette: 0.48798203351890634


In [ ]:
labels = AgglomerativeClustering(n_clusters=3,linkage='single',metric='cityblock').fit_predict(X)
silhouette = silhouette_score(X,labels)
print(f"Distancia: cityblock, Linkage:single, Silhouette: {silhouette}")

Distancia: cityblock, Linkage:single, Silhouette: -0.06929245423622125


In [ ]:
labels = AgglomerativeClustering(n_clusters=3,linkage='average',metric='euclidean').fit_predict(X)
silhouette = silhouette_score(X,labels)
print(f"Distancia: euclidean, Linkage:average, Silhouette: {silhouette}")

Distancia: euclidean, Linkage:average, Silhouette: 0.6100753288756406


In [ ]:
labels = AgglomerativeClustering(n_clusters=3,linkage='average',metric='cityblock').fit_predict(X)
silhouette = silhouette_score(X,labels)
print(f"Distancia: cityblock, Linkage:average, Silhouette: {silhouette}")

Distancia: cityblock, Linkage:average, Silhouette: 0.5840053145196977


In [ ]:
labels = AgglomerativeClustering(n_clusters=3,linkage='complete',metric='euclidean').fit_predict(X)
silhouette = silhouette_score(X,labels)
print(f"Distancia: euclidean, Linkage:complete, Silhouette: {silhouette}")

Distancia: euclidean, Linkage:complete, Silhouette: 0.5418973703348594


In [ ]:
labels = AgglomerativeClustering(n_clusters=3,linkage='complete',metric='cityblock').fit_predict(X)
silhouette = silhouette_score(X,labels)
print(f"Distancia: cityblock, Linkage:complete, Silhouette: {silhouette}")

Distancia: cityblock, Linkage:complete, Silhouette: 0.5418973703348594


In [ ]:
kmeans = KMeans(n_clusters =3,n_init=10,init='random',max_iter=300,random_state=0)
kmeans.fit_predict(X)

In [ ]:
labels = kmeans.labels_
silhouette = silhouette_score(X,labels)
print(f"KMeans silhouette eh {silhouette}")

KMeans silhouette eh 0.5711381937868838


In [ ]:
labels = AgglomerativeClustering(n_clusters=3,linkage='average',metric='euclidean').fit_predict(X)
adjusted_rand = adjusted_rand_score(y,labels)
print(f"Indice rand do melhor eh {adjusted_rand}")

Indice rand do melhor eh 0.292626917173625


Melhor config: Aglomerativo com linkage average e distancia euclidiana, com silhueta média de 0.6100753288756406, rand index de 0.292626917173625

 0.5 < SC < 0.7 significa que uma estrutura razoável foi
encontrada

In [ ]:
dataset = [['Milk', 'Onion', 'Nutmeg', 'Eggs', 'Yogurt'],
           ['Dill', 'Onion', 'Nutmeg', 'Eggs', 'Yogurt'],
           ['Milk', 'Apple', 'Eggs'],
           ['Milk', 'Unicorn', 'Corn', 'Yogurt'],
           ['Corn', 'Onion', 'Onion', 'Ice cream', 'Eggs']]

In [ ]:
dataset

[['Milk', 'Onion', 'Nutmeg', 'Eggs', 'Yogurt'],
 ['Dill', 'Onion', 'Nutmeg', 'Eggs', 'Yogurt'],
 ['Milk', 'Apple', 'Eggs'],
 ['Milk', 'Unicorn', 'Corn', 'Yogurt'],
 ['Corn', 'Onion', 'Onion', 'Ice cream', 'Eggs']]

In [ ]:
te = TransactionEncoder()
te_ary = te.fit(dataset).transform(dataset)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

Apple   Corn   Dill   Eggs  Ice cream   Milk  Nutmeg  Onion  Unicorn  \
0  False  False  False   True      False   True    True   True    False   
1  False  False   True   True      False  False    True   True    False   
2   True  False  False   True      False   True   False  False    False   
3  False   True  False  False      False   True   False  False     True   
4  False   True  False   True       True  False   False   True    False   

   Yogurt  
0    True  
1    True  
2   False  
3    True  
4   False

In [ ]:
frequent_itemsets = apriori(df, min_support=0.5, use_colnames=True)
frequent_itemsets

support       itemsets
0      0.8         (Eggs)
1      0.6         (Milk)
2      0.6        (Onion)
3      0.6       (Yogurt)
4      0.6  (Onion, Eggs)

In [ ]:
df_ar = association_rules(frequent_itemsets, metric = "confidence", min_threshold = 0.7)
df_ar

antecedents consequents  antecedent support  consequent support  support  \
0     (Onion)      (Eggs)                 0.6                 0.8      0.6   
1      (Eggs)     (Onion)                 0.8                 0.6      0.6   

   confidence  lift  leverage  conviction  zhangs_metric  
0        1.00  1.25      0.12         inf            0.5  
1        0.75  1.25      0.12         1.6            1.0

2 regras geradas, (Onion)->(Eggs) e (Eggs)->(Onion), (Onion) é a melhor já que 100% das vezes que ela aparece, eggs também aparece